### DISCUSSION TITLE
Adding a surface to an existing model

### DISCUSSION THREAD

**Discussion by joelmartis** [mjhoptics/ray-optics#141] [Q&A]
I am trying to combine a few optical lenses from Thorlabs, Edmunds etc. I don't see an easy way to simply combine the surfaces in two models (i.e. the thorlabs lens followed by the edmunds lens), so I loaded the first model and decided to manually add the surfaces from the other model. However this turns out to not work:

My code:
---------------------------------------------------------------------------------------------------------

```
#%% I import Thorlabs part https://www.thorlabs.com/thorproduct.cfm?partnumber=PAF2-A4A

from rayoptics.environment import *

root_pth = Path(rayoptics.__file__).resolve().parent

opm, info = open_model("...relevant_directory/TTN136376.ZMX", info=True)

sm  = opm['seq_model']
osp = opm['optical_spec']
pm = `opm['parax_model']`
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']


osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=1)
osp['fov'] = FieldSpec(osp, key=['object', 'height'], value=0, flds=[0]) #point source

sm.gaps[0].thi=3

opm.flip() #flipping the lenses


opm.update_model()


layout_plt = plt.figure(FigureClass=InteractiveLayout, do_draw_rays=True, do_paraxial_layout=False, opt_model=opm)
layout_plt.refresh()

#so far so good!
```

![image](https://github.com/mjhoptics/ray-optics/assets/10430704/1dd1acfa-4c50-4098-b7db-3cc465bbb626)

```
#%% Now I try to manually add the second lens from Edmunds

sm.add_surface([0.064499, 1.6, 'N-BK7', 'Schott'])
sm.add_surface([0.0, 30.0])

sm.list_model() #the output doesn't make sense, the surfaces haven't been added correctly

```
Output:
              c            t        medium     mode   zdr      sd
  Obj:     0.000000      3.00000       air             1      0.0000
    1:     0.127065      1.30000   N-LASF9             1      1.0000
    2:     0.458716      1.11000    N-SK16             1      1.0000
 Stop:    -0.420168      5.00000       air             1      1.0000
    4:    -0.000000      2.80100       air             1     0.90000
  Img:     0.000000      1.60000     N-BK7             1     0.32818
    6:     0.064499      30.0000       air             1      1.0000
    7:     0.000000      0.00000                       1      1.0000

-----------------------------------------------------------------------------------

Your help in correcting this is much appreciated!

Also, once a surface has been added, is there a way to quickly edit it? Such as changing the curvature, thickness or material? And how about being able to add a surface between two existing surfaces?

Thanks,
Joel

**Comment by mjhoptics** [comment]
Hello @joelmartis,
Thanks for your question. I've worked an example below; things to highlight include setting the current surface for the add_surface() command, a workaround to a ray-optics bug that `InteractiveLayout` is missing elements and how to modify the system after it's been entered.
There is also an example in the doc called [Multiple Lens Import Example](https://ray-optics.readthedocs.io/en/latest/examples/AddFromFileAndFlip/AddFromFileAndFlip.html#multiple-lens-import-example)
Please feel free to follow up with any questions.
Regards
Mike Hayford


# Investigate issue #141

```python
from rayoptics.environment import *

opm, info = open_model(Path().parent / "PAF2-A4A-Zemax/TTN136376.ZMX", info=True)

sm  = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']
ar = opm['analysis_results']

osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=1)
osp['fov'] = FieldSpec(osp, key=['object', 'height'], value=0, flds=[0]) #point source

sm.gaps[0].thi=3

opm.flip() #flipping the lenses

opm.update_model()

layout_plt = plt.figure(FigureClass=InteractiveLayout, do_draw_rays=True, do_paraxial_layout=False, opt_model=opm).plot()

#so far so good!
```
![output_7_0](https://github.com/mjhoptics/ray-optics/assets/33294031/2c1cd83a-ea7e-4c96-9909-fe33c5ff1f27)

We need to know the insertion point in the sequence that add_surface will use.

```python
sm.cur_surface
```
    5

```python
sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000      3.00000       air             1      0.0000
        1:     0.127065      1.30000   N-LASF9             1      1.0000
        2:     0.458716      1.11000    N-SK16             1      1.0000
     Stop:    -0.420168      5.00000       air             1      1.0000
        4:    -0.000000      2.80100       air             1     0.90000
      Img:     0.000000      0.00000                       1     0.32818

Surface 5 is the image surface. We want to insert after, say, the dummy plane - surface 4.

```python
sm.set_cur_surface(4)
```

While adding the surfaces, set circular clear apertures to match the Thorlabs OD


```python
sm.add_surface([0.064499, 1.6, 'N-BK7', 'Schott'])
sm.ifcs[sm.cur_surface].clear_apertures.append(srf.Circular(1))

sm.add_surface([0.0, 30.0])
sm.ifcs[sm.cur_surface].clear_apertures.append(srf.Circular(1))

sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000      3.00000       air             1      0.0000
        1:     0.127065      1.30000   N-LASF9             1      1.0000
        2:     0.458716      1.11000    N-SK16             1      1.0000
     Stop:    -0.420168      5.00000       air             1      1.0000
        4:    -0.000000      2.80100       air             1     0.90000
        5:     0.064499      1.60000     N-BK7             1      1.0000
        6:     0.000000      30.0000       air             1      1.0000
      Img:     0.000000      0.00000                       1     0.32818



```python
opm.update_model(src_model=sm)
```

As identified in several other issues, a defect in ray-optics fails to recognize that adding surfaces should add lens elements in the new layout. You have to force ray-optics to rebuild the list of elements it uses for rendering by calling the opm method rebuild_from_seq().


```python
opm.rebuild_from_seq()
```


```python
layout_plt1 = plt.figure(FigureClass=InteractiveLayout, do_draw_rays=True, do_paraxial_layout=False, opt_model=opm).plot()

```

![output_18_0](https://github.com/mjhoptics/ray-optics/assets/33294031/1e4952b3-b1ee-4612-98d6-c42e2b1db1ba)

## create a cemented doublet by inserting a surface between 5 and 6

Insertion using add_surface() inserts the new surface after the cur_surface. This is different from Python list insertion but follows the practice of most image forming software.


```python
sm.set_cur_surface(5)
sm.add_surface([-0.64499,0.9,'N-SF6', 'Schott'], sd=1)

opm.update_model()
opm.rebuild_from_seq()

layout_plt2 = plt.figure(FigureClass=InteractiveLayout, do_draw_rays=True, do_paraxial_layout=False, opt_model=opm).plot()
```

![output_22_0](https://github.com/mjhoptics/ray-optics/assets/33294031/762c55ce-769b-4ad4-b48e-48640961769a)


```python
sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000      3.00000       air             1      0.0000
        1:     0.127065      1.30000   N-LASF9             1      1.0000
        2:     0.458716      1.11000    N-SK16             1      1.0000
     Stop:    -0.420168      5.00000       air             1      1.0000
        4:    -0.000000      2.80100       air             1     0.90000
        5:     0.064499      1.60000     N-BK7             1      1.0000
        6:    -0.644990     0.900000     N-SF6             1     0.39164
        7:     0.000000      30.0000       air             1      1.0000
      Img:     0.000000      0.00000                       1     0.32818


## Editing the sequential model

The individual attributes of the `sm` can be changed directly. It must be followed up by calling update_model() on the `opm`.

The [Sequential Model](https://ray-optics.readthedocs.io/en/latest/models/index.html#sequential-model) is organized around lists of interfaces and gaps. [Interfaces](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.seq.interface.html#rayoptics.seq.interface.Interface) are commonly implemented as Surfaces; [Surfaces](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.surface.html#rayoptics.elem.surface.Surface) support profile definitions and interface with the ray trace. [Gaps](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.seq.gap.html#rayoptics.seq.gap.Gap) consist of a thickness (the separation between adjacent interfaces) and an optical material specification.


```python
sm.gaps[5].medium = create_glass('N-SK16', 'Schott')
sm.gaps[5].thi = 1.2
sm.ifcs[7].profile.cv = -.12

opm.update_model()
opm.rebuild_from_seq()

sm.list_model()
```

                  c            t        medium     mode   zdr      sd
      Obj:     0.000000      3.00000       air             1      0.0000
        1:     0.127065      1.30000   N-LASF9             1      1.0000
        2:     0.458716      1.11000    N-SK16             1      1.0000
     Stop:    -0.420168      5.00000       air             1      1.0000
        4:    -0.000000      2.80100       air             1     0.90000
        5:     0.064499      1.20000    N-SK16             1      1.0000
        6:    -0.644990     0.900000     N-SF6             1     0.39570
        7:    -0.120000      30.0000       air             1      1.0000
      Img:     0.000000      0.00000                       1     0.32818



```python
layout_plt3 = plt.figure(FigureClass=InteractiveLayout, do_draw_rays=True, do_paraxial_layout=False, opt_model=opm).plot()
```

![output_27_0](https://github.com/mjhoptics/ray-optics/assets/33294031/943932d1-a1cd-4258-80f3-88100bf088b4)

**Comment by joelmartis** [comment]
Thank you Michael! This is very helpful!

I had another, not related to this particular issue.

Can I export a ray trace as a wavefront? I want to do some diffraction limited spot size simulations, and exporting the ray trace as a wavefront (i.e. angle v/s phase or something equivalent) would be very helpful.

Joel

**Reply by mjhoptics** [reply]
Hi Joel,
I have a couple of jupyter notebooks that demonstrate wavefront calculations. They are in my notebook repo.

https://github.com/mjhoptics/ray-optics-notebooks/focus%20tilt%20notebook.ipynb
https://github.com/mjhoptics/ray-optics-notebooks/wavefront%20notebook.ipynb

Both of these notebooks setup interactive widgets to change defocus and image position to demonstrate their effect on the wavefront.
Hope these help.
Mike

**Comment by joelmartis** [comment]
Thank you so much Michael!

This is very helpful

-Joel

**Comment by joelfrancis1122** [comment]
yeah it is

**Comment by ebflagg** [comment]
This was very helpful because the existing documentation does not describe how to modify models after they have been created. However, when I followed the example above, I found that the layout plot was not rendering correctly until I moved the `rebuild_from_seq()` method to run *before* the `update_model()` method. I'm using rayoptics==0.9.2, which is the current version as of this date (11/15/24).

**Reply by mjhoptics** [reply]
Hello @ebflagg,
I think the script above will work without any calls to `rebuild_from_seq()`. Version 0.9 had major revisions to keep the sequential model and element model in sync. Everything should be handled by the update_model() call. 
Hope this helps.
Mike Hayford

**Reply by ebflagg** [reply]
@mjhoptics 
You are correct! I had included `rebuild_from_seq()` because it was in the example above. But now I've removed it entirely and the layout plots are correct just from running `update_model()` after modifying the sequential model. In fact, if I run `rebuild_from_seq()` without then running `update_model()` the layout plot becomes incorrect again. (Later-added lenses appear overlapping earlier-added lenses.) So, something might still be wrong with `rebuild_from_seq()`. But since `update_model()` takes care of the updating, we don't really need `rebuild_from_seq()`---at least in the notebook I wrote to practice modifying lenses.
